bounds.h

```diff
@@ -0,0 +1,209 @@
+#ifndef NDARRAY_BOUNDS_H
+#define NDARRAY_BOUNDS_H
+
+#include <tuple>
+
+#include "array.h"
+
+namespace nda {
+
+namespace internal {
+
+template <class IntervalDst, class IntervalSrc>
+NDARRAY_HOST_DEVICE void assert_interval_compatible(size_t interval_index, const IntervalSrc& src) {
+  bool compatible = true;
+  if (is_static(IntervalDst::Min) && !is_dynamic(src.min()) && src.min() != IntervalDst::Min) {
+    NDARRAY_PRINT_ERR("Error converting interval %zu: expected static min " NDARRAY_INDEX_T_FMT
+                      ", got " NDARRAY_INDEX_T_FMT "\n",
+        interval_index, IntervalDst::Min, src.min());
+    compatible = false;
+  }
+  if (is_static(IntervalDst::Extent) && !is_dynamic(src.extent()) &&
+      src.extent() != IntervalDst::Extent) {
+    NDARRAY_PRINT_ERR("Error converting interval %zu: expected static extent " NDARRAY_INDEX_T_FMT
+                      ", got " NDARRAY_INDEX_T_FMT "\n",
+        interval_index, IntervalDst::Extent, src.extent());
+    compatible = false;
+  }
+  assert(compatible);
+  (void)compatible;
+}
+
+template <class IntervalDst, class IntervalSrc, size_t... Is>
+NDARRAY_HOST_DEVICE void assert_intervals_compatible(
+    const IntervalSrc& src, index_sequence<Is...>) {
+  // This is ugly, in C++17, we'd use a fold expression over the comma operator (f(), ...).
+  int unused[] = {(assert_interval_compatible<typename std::tuple_element<Is, IntervalDst>::type>(
+                       Is, nda::interval<>(std::get<Is>(src))),
+      0)...};
+  (void)unused;
+}
+
+template <class IntervalDst, class IntervalSrc>
+NDARRAY_HOST_DEVICE const IntervalSrc& assert_intervals_compatible(const IntervalSrc& src) {
+#ifndef NDEBUG
+  assert_intervals_compatible<IntervalDst>(
+      src, make_index_sequence<std::tuple_size<IntervalDst>::value>());
+#endif
+  return src;
+}
+
+template <class... Intervals, size_t... Is>
+NDARRAY_HOST_DEVICE auto dims(const std::tuple<Intervals...>& intervals, index_sequence<Is...>) {
+  return std::make_tuple(dim(std::get<Is>(intervals))...);
+}
+
+} // namespace internal
+
+template <class... Intervals>
+class bounds;
+
+/** Helper function to make a bounds from a variadic list of `intervals...`. */
+template <class... Intervals>
+NDARRAY_HOST_DEVICE auto make_bounds(Intervals... intervals) {
+  return bounds<Intervals...>(intervals...);
+}
+
+template <class... Intervals>
+NDARRAY_HOST_DEVICE bounds<Intervals...> make_bounds_from_tuple(
+    const std::tuple<Intervals...>& intervals) {
+  return bounds<Intervals...>(intervals);
+}
+
+template <class... Intervals>
```

Comment:
Mostly cribbed from `shape`. I can add more comments once we settle on a design.